### PR TITLE
state total metric

### PR DESCRIPTION
**Pull request by rudigiesler**

### CODE REVIEW

**Comment by rudigiesler** [comment]
@hodgestar ready for review.

File: lib/metrics-helper.js [hunk 1]
```diff
@@ -46,6 +46,46 @@ var MetricsHelper = Eventable.extend(function(self, im) {
...
+
+            Adds an inc metric that fires every time the specified state is
+            entered.
+
+            :param string state:
+                The name of the state to count entries for
+            :param string label:
+                The label for the metric. Defaults to ``total_state_entries``
+        */
+        label = typeof label !== 'undefined' ? label: 'total_state_entries';
```
Comment (hodgestar) [inline lib/metrics-helper.js]: The default label should probably depend on the state name in some way? Otherwise calling this on multiple states will result in the same metric being incremented multiple times?

Comment (rudigiesler) [inline lib/metrics-helper.js]: How should I indicate this in the docs? Is the bash like syntax `total_state_entries_$STATE` clear enough?

Comment (hodgestar) [inline lib/metrics-helper.js]: I think that syntax is good. I'm wondering if we should allow a default here though? State names are potentially any valid string so they could contain really odd things that one wouldn't want in a metric name and I don't know if there is necessarily a sane way to tidy them up. E.g. `state:report` should probably become `enter_state_report`. Thoughts?

Comment (rudigiesler) [inline lib/metrics-helper.js]: Is it a problem to have strange characters in the metric label? would it be bad to have `enter_state:report`?

Comment (hodgestar) [inline lib/metrics-helper.js]: The metrics resource is fairly paranoid about the metric names it accepts
(see
https://github.com/praekelt/vumi-go/blob/develop/go/apps/jsbox/metrics.py#L27
for the regex). Perhaps overly paranoid. :)

Comment (rudigiesler) [inline lib/metrics-helper.js]: So maybe we include the state name in metric label, but replace anything that is not `a-zA-Z0-9._` with `_`? Or would you rather leave the default out completely?

Comment (hodgestar) [inline lib/metrics-helper.js]: Let's go with the replacing -- if people don't like the names we generate
they can always override it. :)

**Comment by rudigiesler** [comment]
@hodgestar I made changes to how the metric works as per the Slack discussion. Ready for review again.

File: lib/metrics-helper.js [hunk 2]
```diff
@@ -46,6 +46,52 @@ var MetricsHelper = Eventable.extend(function(self, im) {
         return self;
     };
 
+    self.add.total_state_actions = function(state, label) {
+        /** MetricsHelper.add.total_state_entries(label)
```
Comment (hodgestar) [inline lib/metrics-helper.js]: The docstring function name and arguments need updating.

File: lib/metrics-helper.js [hunk 3]
```diff
@@ -46,6 +46,52 @@ var MetricsHelper = Eventable.extend(function(self, im) {
...
+        */
+        var action = (typeof state.action === 'undefined' ? '' : state.action)
+            .match(/(enter|exit|input|resume|show)/i);
+        action = action === null ? 'enter' : action[0].toLowerCase();
+
+        label = typeof label !== 'undefined' 
+            ? label
+            : ['total_action', 
+               action, 
+               state.state.replace(/[^a-zA-Z._]/g, '_')].join('_');
```
Comment (hodgestar) [inline lib/metrics-helper.js:89]: I'm guessing we're going to want tiny helper functions for parsing state event names and translating labels. Perhaps we should do that refactoring in a separate PR though (or maybe just the next PR when we need them again).

File: lib/metrics-helper.js [hunk 4]
```diff
@@ -46,6 +46,52 @@ var MetricsHelper = Eventable.extend(function(self, im) {
...
+                Defaults to ``enter``.
+            :param string label:
+                The label for the metric. Defaults to 
+                ``total_action_$ACTION_$STATE`` where ``$ACTION`` is the
+                specified action, and ``$STATE`` is the name of the state with
+                all characters that are not ``a-zA-Z._`` replaced with ``_``.
+        */
+        var action = (typeof state.action === 'undefined' ? '' : state.action)
+            .match(/(enter|exit|input|resume|show)/i);
+        action = action === null ? 'enter' : action[0].toLowerCase();
```
Comment (hodgestar) [inline lib/metrics-helper.js]: This translates any unknown action into `enter` but I think it'd be better to rather throw an error?

**Comment by hodgestar** [comment]
Minor comments. Looking good.

File: lib/metrics-helper.js (hunk 3 above)
Comment (rudigiesler) [inline lib/metrics-helper.js:89]: I'll do it in the next PR that I find that I need these functions again.

**Comment by rudigiesler** [comment]
@hodgestar Made the changes. Ready for review again.

File: lib/metrics-helper.js [hunk 5]
```diff
@@ -46,6 +46,56 @@ var MetricsHelper = Eventable.extend(function(self, im) {
...
+                ``total_action_$ACTION_$STATE`` where ``$ACTION`` is the
+                specified action, and ``$STATE`` is the name of the state with
+                all characters that are not ``a-zA-Z._`` replaced with ``_``.
+        */
+        var action = (typeof state.action === 'undefined'?'enter':state.action)
+            .match(/(enter|exit|input|resume|show)/i);
+        try {
+            action = action[0].toLowerCase();
+        } catch(e) {
+            throw "Invalid state action " + state.action;
```
Comment (hodgestar) [inline lib/metrics-helper.js:82]: We should probably throw an exception object rather than a string here. The toolkit has a base class for this. See https://github.com/praekelt/vumi-jssandbox-toolkit/blob/master/lib/interaction_machine.js#L44 for how it's used.

Comment (hodgestar) [inline lib/metrics-helper.js:82]: Maybe let's do a nicer-errors PR so this can land though?

**Comment by hodgestar** [comment]
:+1:

File: lib/metrics-helper.js (hunk 5 above)
Comment (rudigiesler) [inline lib/metrics-helper.js:82]: Created issue: https://github.com/praekelt/go-jsbox-metrics-helper/issues/7
